core/src/abstractions.rs

```diff
@@ -43,3 +56,84 @@ impl MeteredSemaphore {
         }
     }
 }
+
+#[derive(Clone)]
+pub(crate) struct StreamAllowHandle {
+    acceptable_to_get_next: Arc<AtomicBool>,
```

Comment:
I'm pretty sure you can implement this without the atomic bool if you call `notify_one` on construction.

Reply:
Yeah good point